### PR TITLE
Move clock to sun sensors

### PR DESCRIPTION
**Pull request by oxan**
# What does this implement/fix? 

Move ownership of the clock from the `Sun` class ot the sensors, so the `Sun` class can be used more generically.

Small breaking change because with it the `time_id` option moved from the `sun` component to the sensors as well.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [x] ESP8266

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by RebbePod** [comment]
Is it better to pass to the sun a `time::ESPTime` over a timestamp `time_t`?

**Comment by oxan** [comment]
> Is it better to pass to the sun a `time::ESPTime` over a timestamp `time_t`?

That's a good question. Let's keep using `time::ESPTime` for now, cleaning up time handling is something for another time.

**Comment by RebbePod** [comment]
> That's a good question. Let's keep using `time::ESPTime` for now, cleaning up time handling is something for another time.

Sounds good. I would like to point out that the only part of the esptime that's being used is the timestamp.

**Comment by oxan** [comment]
No, `Sun::elevation()` and `Sun::azimuth()` use the others through `Sun::calc_coords()` which ultimately calls `julian_day()`.

File: esphome/const.py [hunk 1]
```diff
@@ -90,6 +90,7 @@
 CONF_CHIPSET = "chipset"
 CONF_CLIENT_ID = "client_id"
 CONF_CLK_PIN = "clk_pin"
+CONF_CLOCK_ID = "clock_id"
```
Comment (OttoWinter) [inline esphome/const.py:93]: Unused

```suggestion
```

File: esphome/components/sun/sun.h [hunk 2]
```diff
@@ -67,7 +66,7 @@ class Sun {
   internal::HorizontalCoordinate calc_coords_();
   optional<time::ESPTime> calc_event_(bool rising, double zenith);
 
-  time::RealTimeClock *time_;
+  time::ESPTime time_;
```
Comment (OttoWinter) [inline esphome/components/sun/sun.h:69]: Tbh I can't say I like this interface, If I call `sunrise(0)` I will get different results based on what clock most recently called this method. Can't we just pass the `time::ESPTime` to the `sunrise` etc functions directly as parameters?

Comment (RebbePod) [reply esphome/components/sun/sun.h:69]: Is that possible with the current implementation?

Comment (OttoWinter) [reply esphome/components/sun/sun.h:69]: I don't see why it wouldn't be... The only method of `time_` that is accessed in the original version is `utcnow()`, which we could also just supply as a parameter

**Comment by RebbePod** [comment]
Where is this holding?

**Comment by OttoWinter** [comment]
> Where is this holding?

See the last two review comments (https://github.com/esphome/esphome/pull/2933#pullrequestreview-887776402, https://github.com/esphome/esphome/pull/2933#discussion_r810353695)

**Pull request closed without merging**